rpc-broker/src/rpc-broker.c

```diff
@@ -0,0 +1,444 @@
+    printf("Adds extra information (run with logging).\n");
+    printf("\t-w  [--websockets=PORT]                 ");
+    printf("Sets rpc-broker to run on given address/port as websockets.\n");
+}
+
+void sigint_handler(int signal)
+{
+    DBUS_BROKER_WARNING("<received signal interrupt> %s", "");
+    free_dlinks();
+    free_policy();
+    // handle cleanup of the uv handles
+    if (ring)
+        lws_ring_destroy(ring);
+
+    exit(0);
+}
+
+void sighup_handler(int signal)
+{
+    reload_policy = true;
+    DBUS_BROKER_EVENT("Re-loading policy %s", "");
+}
+
+/*
+ * Cycle through linked-list of current dbus-signals being subscribed to.
+ * If there are any signals queued up, start the exchange of communication.
+ */
+static void service_ws_signals(void)
+{
+    struct dbus_link *curr = dlinks;
+
+    while (curr) {
+
+        DBusMessage *msg = NULL;
+
+        if (curr->dconn && dbus_connection_get_is_connected(curr->dconn)) {
+            dbus_connection_read_write(curr->dconn, 0);
+            msg = dbus_connection_pop_message(curr->dconn);
+        }
+
+        if (!msg)
+            goto next_link;
+
+        if (dbus_message_get_type(msg) != DBUS_MESSAGE_TYPE_SIGNAL)
+            goto unref_msg;
+
+        struct json_response *jrsp = init_jrsp();
+        jrsp->response_to[0] = '\0';
+        snprintf(jrsp->type, JSON_REQ_ID_MAX - 1, "%s", JSON_SIG);
+
+        load_json_response(msg, jrsp);
+
+        jrsp->interface = dbus_message_get_interface(msg);
+        jrsp->member = dbus_message_get_member(msg);
+        jrsp->path = dbus_message_get_path(msg);
+
+        char *reply = prepare_json_reply(jrsp);
+
+        if (!reply)
+            goto free_msg;
+
+        lws_callback_on_writable(curr->wsi);
+        lws_ring_insert(ring, reply, 1);
+        free(reply);
+
+free_msg:
+        free(jrsp);
+
+unref_msg:
+        dbus_message_unref(msg);
+
+next_link:
+        curr = curr->next;
+    }
+}
+
+static void run_websockets(struct dbus_broker_args *args)
+{
+    struct lws_context *ws_context = NULL;
+    ws_context = create_ws_context(args->port);
+    if (dom0)
+        dbus_broker_policy = build_policy(args->rule_file);
+
+    if (!ws_context)
+        DBUS_BROKER_ERROR("WebSockets-Server");
+
+    DBUS_BROKER_EVENT("<WebSockets-Server has started listening> [Port: %d]",
+                        args->port);
+
+    while (dbus_broker_running) {
+
+        lws_service(ws_context, WS_LOOP_TIMEOUT);
+        service_ws_signals();
+
+        if (reload_policy) {
+            free_policy();
+            dbus_broker_policy = build_policy(args->rule_file);
+            reload_policy = false;
+        }
+    }
+
+    lws_ring_destroy(ring);
+    lws_context_destroy(ws_context);
+}
+
+static void close_connection(uv_handle_t *handle)
+{
+    struct raw_dbus_conn *rdconn = (struct raw_dbus_conn *) handle->data;
+    free(rdconn);
+}
+
+void service_rdconn_cb(uv_poll_t *handle, int status, int events)
```

Comment:
Since this is only called/referenced by rawdbus below, why not make this static?